merkle.go

```diff
@@ -0,0 +1,189 @@
+				Left: nodes[result.storeIndex],
+				// Save a reference to the right child
+				Right: nodes[result.storeIndex+1],
+			}
+
+			// Save the parent reference with the children
+			nodes[result.storeIndex].Parent = parent
+			nodes[result.storeIndex+1].Parent = parent
+
+			// Overwrite the left child's slot in the array,
+			// since it's no longer needed. The right child
+			// is also not needed anymore in the original array,
+			// and will be overwritten later
+			nodes[result.storeIndex] = parent
+		}
+
+		// Now that results are gathered for the level,
+		// the array can be shifted and shrunk
+		shiftAndShrinkArray(&nodes)
+	}
+
+	return &MerkleTree{
+		Root: nodes[0],
+	}, nil
+}
+
+// isValidDataSet makes sure the input set has elements
+func isValidDataSet(inputData [][]byte) bool {
+	return inputData != nil && len(inputData) > 0
+}
+
+// shiftAndShrinkArray shifts every other node to the
+// beginning of the array, and discards half of it (shrinks it).
+// Due to the way results are being stored (index of left child),
+// and the fact that the Merkle tree is a perfect binary tree,
+// it can be guaranteed that the results are on every other index in the node level array
+func shiftAndShrinkArray(nodes *[]*Node) {
+	// Put the results in the first half of the array.
+	// One counter keeps track of the next slot to place the value (moves by 1) (saveIndx)
+	// and the other keeps track of which element should be stored (moves by 2) (resultIndx)
+	initialLevelSize := len(*nodes)
+	saveIndx := 0
+	for resultIndx := 0; resultIndx < initialLevelSize; resultIndx += 2 {
+		(*nodes)[saveIndx] = (*nodes)[resultIndx]
+		saveIndx++
+	}
+
+	// Wipe the other half of the array, since
+	// all useful and needed results are in the first half
+	*nodes = (*nodes)[:initialLevelSize/2]
+}
+
+// generateLeaves generates the initial (leaf) level of the Merkle tree.
+// The leaf level needs to be a power of 2, since the Merkle tree is considered
+// to be a perfect binary tree
+func generateLeaves(inputData [][]byte, wp *workerPool) ([]*Node, error) {
+	inputDataSize := len(inputData)
+	leafLevelSize := inputDataSize
+
+	if inputDataSize > 1 {
+		// Find the nearest power of 2 for the base leaf level size
+		leafLevelSize = nearestPowerOf2(len(inputData))
+	}
+
+	leaves := make([]*Node, leafLevelSize)
+
+	// Create the initial job set for the leaf nodes,
+	// where each job is a single leaf node to be processed
+	for i := 0; i < inputDataSize; i++ {
+		wp.addJob(&workerJob{
+			storeIndex: i,
+			sourceData: [][]byte{
+				inputData[i],
+			},
+		})
+	}
+
+	// Grab the results from the worker pool
+	for i := 0; i < inputDataSize; i++ {
+		result := wp.getResult()
+		if result.error != nil {
+			return nil, fmt.Errorf(
+				"unable to perform hashing, %w",
+				result.error,
+			)
+		}
+
+		// Save the leaf nodes
+		leaves[i] = &Node{
+			Hash:   result.hashData,
+			Left:   nil,
+			Right:  nil,
+			Parent: nil,
+		}
+	}
```

Comment:
I'm little bit confusing because we have guarantee that all jobs finishes and return in same order?

Reply:
`storeIndex` can be used?

Reply:
@zivkovicmilos No, I meant why you can assume all data at same level come in order? For example, in GenerateMerkleTree, you use storeIndex to decide where you set the node like this:
because you don't know which node comes first.

```go
nodes[result.storeIndex] = parent
```

But in generateLeaves, you don't. use storeIndex. Why you can set like this?

```go
leaves[i] = &Node{
			Hash:   result.hashData,
			Left:   nil,
			Right:  nil,
			Parent: nil,
		}
```

Reply:
You don't control the order, you always place the parent at the slot of the left child:
https://github.com/Trapesys/fastmerkle/blob/7cb1739221c304730ae9cdd4a8b16732efa01961/merkle.go#L63-L81

Reply:
Leaves can be in random order?

Reply:
Ah, yes, because you override them anyways (the parent node is the one that's important since you're saving a reference to them when you create the the parent), and their positions in the array are useless since the array is being snipped on every level

Reply:
I'm an idiot 🤦 

For the entire time I'm looking at this code snippet like it's from the main parse loop, and not the leaves loop.

You're completely right, we should use `storeIndex` here 👍 

Fixed:

[c12d713](https://github.com/Trapesys/fastmerkle/pull/1/commits/c12d7131f0d7469c29c1e42bc0340a9a16fd77e8)